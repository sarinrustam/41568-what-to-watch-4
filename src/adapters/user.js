export const userAdapter = (data) => {
  return {
    avatarUrl: data[`avatar_url`],
  };
};
