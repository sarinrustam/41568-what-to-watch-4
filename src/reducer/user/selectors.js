import NameSpace from "../name-space.js";

export const getAuthStatus = (state) => {
  return state[NameSpace.USER].authorizationStatus;
};

export const getAvatar = (state) => {
  return state[NameSpace.USER].avatar;
};

export const getErrorMessage = (state) => {
  return state[NameSpace.USER].authorizationError;
};
