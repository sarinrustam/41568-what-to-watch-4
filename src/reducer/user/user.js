import {extend} from "../../utils/utils.js";
import {userAdapter} from "../../adapters/user.js";

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const defaultAvatar = `img/avatar.jpg`;

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  avatar: defaultAvatar,
  authorizationError: ``,
  checkAuthIsLoaded: false
};

const ActionType = {
  SET_AUTHORIZATION: `SET_AUTHORIZATION`,
  ADD_AVATAR: `ADD_AVATAR`,
  ADD_AUTH_ERROR: `ADD_AUTH_ERROR`,
  SET_CHECK_AUTH_IS_LOADED: `SET_CHECK_AUTH_IS_LOADED`
};

const ActionCreator = {
  setAuthorization: (status) => {
    return {
      type: ActionType.SET_AUTHORIZATION,
      payload: status,
    };
  },
  addAvatar: (srcImage) => {
    return {
      type: ActionType.ADD_AVATAR,
      payload: srcImage,
    };
  },
  addAuthError: (errorMessage) => {
    return {
      type: ActionType.ADD_AUTH_ERROR,
      payload: errorMessage,
    };
  },
  setCheckAuthIsLoaded: (value) => {
    return {
      type: ActionType.SET_CHECK_AUTH_IS_LOADED,
      payload: value
    };
  }
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then(({data}) => {
        dispatch(ActionCreator.setAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.addAvatar(userAdapter(data).avatarUrl));
        dispatch(ActionCreator.setCheckAuthIsLoaded(true));
      })
      .catch(() => {
        dispatch(ActionCreator.setCheckAuthIsLoaded(true));
      });
  },
  login: ({email, password}) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email,
      password
    })
      .then(({data}) => {
        dispatch(ActionCreator.setAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.addAvatar(userAdapter(data).avatarUrl));
      })
      .catch(({response}) => {
        dispatch(ActionCreator.addAuthError(response.data.error));
      });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload,
      });
    case ActionType.ADD_AVATAR:
      return extend(state, {
        avatar: action.payload,
      });
    case ActionType.ADD_AUTH_ERROR:
      return extend(state, {
        authorizationError: action.payload,
      });
    case ActionType.SET_CHECK_AUTH_IS_LOADED:
      return extend(state, {
        checkAuthIsLoaded: action.payload,
      });
    default:
      return state;
  }
};

export {reducer, ActionType, ActionCreator, Operation, AuthorizationStatus};
