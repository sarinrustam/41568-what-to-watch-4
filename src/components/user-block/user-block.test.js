import React from "react";
import rerender from "react-test-renderer";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import NameSpace from "../../reducer/name-space.js";
import {AuthorizationStatus} from "../../reducer/user/user.js";

import UserBlock from "./user-block.jsx";

const mockStore = configureStore([]);

describe(`UserBlock test`, () => {
  it(`SignIn renfer correctly`, () => {
    const store = mockStore({
      [NameSpace.USER]: {
        authorizationStatus: AuthorizationStatus.NO_AUTH,
        avatar: `s`,
        authorizationError: ``,
      }
    });
    const tree = rerender.create(
        <Provider store={store}>
          <UserBlock
          />
        </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
