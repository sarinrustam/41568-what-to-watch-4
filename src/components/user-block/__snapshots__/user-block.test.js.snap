// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserBlock test SignIn renfer correctly 1`] = `
<div
  className="user-block"
>
  <a
    className="user-block__link"
    href="/login"
  >
    Sign in
  </a>
</div>
`;
